### ISSUE TITLE
a float32 value in a structure is stored incorrectly above a given number

### ISSUE THREAD

**Issue by dRacz3** [apache/arrow#37956]
### Describe the bug, including details regarding any error messages, version, and platform.

I am trying to write a structured data using the following schema. 
Using pyarrow==13.0.0
on a MacOS Monterey 12.1, M1 macbook

It seems like, that above a certain value, a floating point number is written incorrectly to the resulting parquet file. The error seems to depend on the actual value of the number.

This is the minimum code where I could replicate this behaviour:
```python
import pandas as pd
import pyarrow as pa

schema = pa.schema(
    [
        pa.field(
            name="root",
            type=pa.struct(
                [
                    pa.field(name="timestamp", type=pa.float32(), nullable=True),
                ]
            ),
            nullable=True,
        )
    ]
)

def test_pyarrow_write_bug():

    data = [
        {'root': {'timestamp': 1.0}},
        {'root': {'timestamp': 1693260000.0}},
        {'root': {'timestamp': 846630000.0}},
        {'root': {'timestamp': 423315000.0}},
        {'root': {'timestamp': 211657500.0}},
        {'root': {'timestamp': 105828750.0}},
        {'root': {'timestamp': 52914375.0}},
        {'root': {'timestamp': 26457187.0}},
    ]
    df = pd.DataFrame(data)
    table = pa.Table.from_pandas(df, schema=schema)
    pa.parquet.write_table(table, 'test_pyarrow_write_bug.parquet')

    df_read = pd.read_parquet('test_pyarrow_write_bug.parquet')
    assert df['root'].to_list() == df_read['root'].to_list()
```
<img width="895" alt="image" src="https://github.com/apache/arrow/assets/21102313/4516e684-b686-442b-9e94-7f1d1dd8bf64">

I would expect the written and read files 

### Component(s)

Python

**Comment by jorisvandenbossche** [comment]
@dRacz3 Thanks for the report! 
This actually already happens in the initial conversion from python objects to pyarrow Table (and so not in the writing to Parquet). Using an even simpler example to illustrate:

```
>>> pa.array([1.0, 1693260000.0, 26457187.0], pa.float32())
<pyarrow.lib.FloatArray object at 0x7f543a06c880>
[
  1,
  1693260000,
  26457188
]
```

So the last value here is "converted" from ending on 87 to ending on 88. I assume this is simply due to the limited precision of float32 (compared to float64) and it cannot represent any number. 

Note that this is not unique to pyarrow. For example also numpy shows the same behaviour:

```
>>> np.array([26457187.0], np.float32())[0]
26457188.0
```

I would have expected us to throw an error, though. Especially when explicitly casting from float64 to float32:

```
>>> pa.array([1.0, 1693260000.0, 26457187.0], pa.float64())
<pyarrow.lib.DoubleArray object at 0x7f5440614e80>
[
  1,
  1693260000,
  26457187
]

>>> pa.array([1.0, 1693260000.0, 26457187.0], pa.float64()).cast(pa.float32())
<pyarrow.lib.FloatArray object at 0x7f543a0a3d00>
[
  1,
  1693260000,
  26457188
]
```

**Comment by dRacz3** [comment]
Thanks for the quick reply, however it does not really explain to me how does 1693260000.0 become 16932600032.0?

**Comment by pitrou** [comment]
Yeah, there is something weird in how both Python and Numpy represent float32 values. I'm not sure where that comes from.

**Comment by pitrou** [comment]
Ok, I've pinned it down to our use of the [double-conversion](https://github.com/google/double-conversion) library.

Note that the output is correct, as it does roundtrip to the actual value when converted back to float32. It's just unintuitive.

In more detail, this is because we are using the `ToShortest` API documented below:
https://github.com/google/double-conversion/blob/a23f2c05376bec668e108a2edd16c83c2618ea52/double-conversion/double-to-string.h#L202-L247

If we want to fix this, we will have to change our underlying float formatting library, but that can come with its own set of downsides.
